src/polymer-ide.ts

```diff
@@ -27,68 +27,96 @@ import TooltipManager from './tooltip-manager';
+  private getProjectPathFromEditor(editor: AtomCore.IEditor) {
+    return atom.project.relativizePath(editor.getPath())[0];
   };
 
-  setProjectPaths(projectPaths: string[]) {
-    if (projectPaths.length === 0) {
-      this.linter.configurationError = 'polymer-ide only works with projects.';
-    } else if (projectPaths.length > 1) {
-      this.linter.configurationError =
-          `polymer-ide only works with projects with exactly one root ` +
-          `directory, this project has: ${JSON.stringify(projectPaths)}`;
+  private updateWorkspaceMap(editor: AtomCore.IEditor) {
+    const project = this.getProjectPathFromEditor(editor);
+    editor.onDidDestroy(() => {
+      const workspace = this.workspaces.get(project);
+      workspace[0] = workspace[0] - 1;
+    });
+    if (this.workspaces.has(project)) {
+      this.workspaces.get(project)[0] = this.workspaces.get(project)[0] + 1;
     } else {
-      const rootDir = projectPaths[0];
-      if (this.editorService) {
-        this.editorService.dispose();
+      // TODO(timvdlippe): Expose as user setting?
+      if (this.workspaces.size == 10) {
+        for (const oldProject of this.workspaces.keys()) {
+          const workspace = this.workspaces.get(oldProject);
+          if (workspace[0] <= 1) {
+            this.workspaces.delete(oldProject);
+            break;
+          }
+        }
       }
-      this.editorService = new RemoteEditorService(rootDir);
-      this.subscriptions.add(this.editorService);
-      this.linter.configurationError = null;
-      this.linter.editorService = this.editorService;
-      this.autocompleter.editorService = this.editorService;
+      const editorService = new RemoteEditorService(project);
+      const workspace = {
+        linter: new Linter(editorService),
+        autocompleter: new Autocompleter(editorService),
+        editorService: editorService,
+        tooltipManager: new TooltipManager(editorService)
+      };
+      this.workspaces.set(project, [1, workspace]);
     }
   };
 
-  serialize(): ViewState {
-    return {};
+  provideLinter(): lint.Provider {
+    return {
+      name: 'polymer-ide',
+      grammarScopes: ['source.js', 'text.html', 'text.html.basic'],
+      scope: 'file',
+      lintOnFly: true,
+      lint: (editor) => {
+        const project = this.getProjectPathFromEditor(editor);
+        return this.workspaces.get(project)[1].linter.lint(editor);
+      }
+    }
   };
 
-  provideLinter(): lint.Provider|lint.Provider[] {
-    return this.linter;
+  provideAutocompleter(): autocomplete.Provider {
+    return {
+      selector: '.text.html, .source.js',
+      getSuggestions: (options) => {
+        const project = this.getProjectPathFromEditor(options.editor);
+        return this.workspaces.get(project)[1].autocompleter.getSuggestions(options);
+      }
+    };
+  };
+
+  deactivate() {
+    this.subscriptions.dispose();
+    this.subscriptions = null;
+    this.workspaces = new Map();
```

Comment:
Do we also need to clean up the linter, autocompleter and tooltipmanager or is this already sufficient? We could also add them to `this.subscriptions` when we create them inside `updateWorkspaceMap`